File: 70. Climbing Stairs.java
import java.util.HashMap;

class Solution {
    HashMap<Integer,Integer> cache = new HashMap<>();
    public int climbStairs(int n) {
        if(n==0){
            return 1;
        } else  if(n<0){
            return 0;
        } else if(cache.containsKey(n)){
            return cache.get(n);
        }
        cache.put(n,climbStairs(n-1) + climbStairs(n-2));
        return cache.get(n);
    }
}

File: 70. Climbing Stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        cache = dict()
        def dfs(n: int):
            if n < 0:
                return 0
            elif n == 0:
                return 1
            elif n in cache:
                return cache[n]
            cache[n] = dfs(n-1) + dfs(n-2)
            return cache[n]
        return dfs(n)
    
    def climbStairsSemp2023(self, n: int) -> int:

        result, cache = 0, dict() 

        def dfs(step) -> int:
            if step == n: return 1
            elif step > n: return 0 
            elif step in cache:  return cache[step]
            cache[step] = dfs(step+1) + dfs(step+2)
            return cache[step]            

        return dfs(0)